OrbitQt/servicedeploymanager.h

```diff
@@ -35,6 +36,13 @@ class ServiceDeployManager : public QObject {
       const GrpcPort& grpc_port, QObject* parent = nullptr);
 
   outcome::result<GrpcPort> Exec();
+
+  // This method opens SftpChannel and initiates a asynchronous copy of the
+  // remote source file to destination.
+  void CopyFileToLocal(std::string_view source, std::string_view destination,
+                       std::function<void()> copy_complete_callback,
+                       std::function<void(ErrorMessage)> error_callback);
```

Comment:
I see why you're hesitant about callbacks. We have callbacks, inside callbacks, inside callbacks here.

One option to improve that would be not to provide a function here taking two callbacks. I would rather use `SftpCopyToLocalOperation` directly where it's actually used. But that's in OrbitGl/App.cpp, isn't it?

Do you see a sensible way to move that to OrbitQt?

If not, I might have some other idea around QFuture/std::future but I first have to look into it.

Reply:
Yes, I see. I guess this one-line should be blocking and it should be callable from another thread?

Reply:
Yes, that would be the idea. Having a thread with synchronous calls reduces number of callbacks to 1 (usually), which makes code more readable. Also it reduces the scope of some local variables in case you need to do more than one thing.

These are 2 main reasons I do not like asynchronous way of writing the code.

Reply:
I see. And I kind of agree.

What about a wrapper function which can be called from any thread? Something like this?

```
#include "servicedeploymanager.h"
#include <future>
#include <QMetaObject>
#include <QCoreApplication>

ErrorMessageOr<void> copyFileFromRemote(ServiceDeployManager* mgr, std::string source, std::string dest) {

    std::promise<ErrorMessageOr<void>> promise;
    auto future = promise.get_future();

    // That queues the lambda-execution on the main thread.
    QMetaObject::invokeMethod(QCoreApplication::instance(),
      [promise = std::move(promise), mgr, source, dest]() {
        // This assume the previous (blocking) implementation of CopyFileToLocal
        promise.set_result(mgr.CopyFileToLocal(source, dest));
    });

    // This waits for the result
    return future.get();
}
```

Reply:
done - in a slightly different way.